### PR TITLE
fix(history-service): navigation after reload

### LINKED ISSUE #171: History Service: Reload bug
Navigating forward, then reloading the page and trying to navigate backwards does not work.

### PR DESCRIPTION
**Pull request by KingHenne**
fixes #171

Instead of using a memory history under the hood for the consumer history, and storing the entries in session storage, we just compute the consumer path from the root location, and store all consumer states in the root history state.

This change also removes the memory history. We don't need a facade for memory histories, since they don't affect the window location and browser history. For server-side rendering, we provide a static history instead.

### CODE REVIEW

File: packages/history-service/src/__tests__/index.test.ts [hunk 1]
```diff
@@ -514,19 +528,27 @@ describe('defineHistoryService', () => {
...
-            history1.listen(history1ListenerSpy);
+            history1Unregister = history1.listen(history1ListenerSpy);
 
             history2ListenerSpy = jest.fn();
-            history2.listen(history2ListenerSpy);
-          });
+            history2Unregister = history2.listen(history2ListenerSpy);
+          };
+
+          const unregisterHistories = () => {
```
Comment (clebert) [inline packages/history-service/src/__tests__/index.test.ts]: maybe rename it to destroyHistories analogous to createHistories

File: packages/history-service/src/base-history.ts [hunk 2]
```diff
@@ -5,28 +5,20 @@ export interface ConsumerHistory extends history.History {
   destroy(): void;
 }
 
-function setRootHistoryKey(
-  consumerLocation: history.Location,
-  rootHistoryKey: string | undefined
-): void {
-  consumerLocation.state = {...consumerLocation.state, rootHistoryKey};
+export interface ConsumerHistoryIndices {
+  [consumerId: string]: number | undefined;
```
Comment (clebert) [inline packages/history-service/src/base-history.ts]: readonly?

File: packages/history-service/src/browser-history.ts [hunk 3]
```diff
@@ -1,19 +1,37 @@
...
+  ) {
+    super(consumerId, rootHistory, rootLocationTransformer);
+
+    const index = this.getConsumerHistoryIndexFromRootHistory();
+
+    if (index >= 0) {
+      this.restoreConsumerHistoryEntries();
+    } else {
+      this.persistConsumerLocation(this.consumerHistory.location, 'replace');
+      this.consumerHistory.action = 'POP';
```
Comment (clebert) [inline packages/history-service/src/browser-history.ts]: newline?

File: packages/history-service/src/history-service-v1.ts [hunk 4]
```diff
@@ -2,13 +2,10 @@ import * as history from 'history';
...
-import {MemoryHistory} from './memory-history';
 import {RootHistories} from './root-histories';
 
 export interface HistoryServiceV1 {
-  readonly rootLocation?: history.Location;
   createBrowserHistory(): history.History;
-  createMemoryHistory(): history.MemoryHistory;
 }
 
 export class HistoryService implements HistoryServiceV1 {
```
Comment (clebert) [inline packages/history-service/src/history-service-v1.ts]: i think this name does not scale in future :)

File: packages/history-service/src/history-service-v1.ts [hunk 5]
```diff
@@ -19,10 +16,6 @@ export class HistoryService implements HistoryServiceV1 {
     private readonly consumerHistories: ConsumerHistory[]
```
[the remark this answers is not in the document]
Comment (clebert) [reply packages/history-service/src/history-service-v1.ts]: consumerId should be the first argument, like in BrowserHistory

File: packages/history-service/src/base-history.ts [hunk 6]
```diff
@@ -5,70 +5,54 @@ export interface ConsumerHistory extends history.History {
...
-  }
-
-  public get length(): number {
-    return this.consumerHistory.length;
   }
 
-  public get action(): history.Action {
-    return this.consumerHistory.action;
-  }
+  public abstract get length(): number;
```
Comment (clebert) [inline packages/history-service/src/base-history.ts]: maybe group the abstract stuff together

File: packages/history-service/src/base-history.ts [hunk 7]
```diff
@@ -5,70 +5,54 @@ export interface ConsumerHistory extends history.History {
...
-function setRootHistoryKey(
-  consumerLocation: history.Location,
-  rootHistoryKey: string | undefined
-): void {
-  consumerLocation.state = {...consumerLocation.state, rootHistoryKey};
+export interface ConsumerHistoryStates {
+  readonly [consumerId: string]: unknown;
 }
 
 export abstract class BaseHistory implements ConsumerHistory {
```
Comment (clebert) [inline packages/history-service/src/base-history.ts]: at this point in time we do not need this base class (yagni ;)

File: packages/history-service/src/history-service-v1.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (KingHenne) [reply packages/history-service/src/history-service-v1.ts]: fixed

File: packages/history-service/src/history-service-v1.ts (hunk 4 above)
Comment (KingHenne) [reply packages/history-service/src/history-service-v1.ts]: fixed by refactoring to a `HistoryServiceV1Binding` class

File: packages/history-service/src/base-history.ts (hunk 6 above)
Comment (KingHenne) [reply packages/history-service/src/base-history.ts]: done

File: packages/history-service/src/base-history.ts (hunk 7 above)
Comment (KingHenne) [reply packages/history-service/src/base-history.ts]: as discussed, I will add the static history in this PR

File: packages/history-service/src/root-histories.ts [hunk 8]
```diff
@@ -3,42 +3,20 @@ import * as history from 'history';
 
 export interface RootHistories {
   readonly browserHistory: history.History;
-  readonly memoryHistory: history.MemoryHistory;
 }
 
 export function createRootHistories(
-  serverRequest?: ServerRequest
+  _serverRequest?: ServerRequest // later needed for static root history
```
[the remark this answers is not in the document]
Comment (KingHenne) [reply packages/history-service/src/root-histories.ts]: as discussed, I will add the static history in this PR

File: packages/history-service/src/history-service-v1.ts [hunk 9]
```diff
@@ -1,53 +1,73 @@
...
 import {RootLocationTransformer} from '.';
-import {ConsumerHistory} from './base-history';
 import {BrowserHistory} from './browser-history';
-import {MemoryHistory} from './memory-history';
 import {RootHistories} from './root-histories';
+import {StaticHistory} from './static-history';
 
 export interface HistoryServiceV1 {
-  readonly rootLocation?: history.Location;
+  staticRootLocation: history.Location;
```
Comment (clebert) [inline packages/history-service/src/history-service-v1.ts]: maybe we should rename replace the `root` prefix with `shared` everywhere

File: packages/history-service/src/static-history.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import {ServerRequest} from '@feature-hub/server-renderer';
+import * as history from 'history';
+import {BaseHistory} from './base-history';
+import {ConsumerHistoryStates} from './root-histories';
+
+/* istanbul ignore next */
+const noop = () => undefined;
+
+export class StaticRootHistory
```
Comment (clebert) [inline packages/history-service/src/static-history.ts]: Please move this to its own file

File: packages/history-service/src/base-history.ts [hunk 11]
```diff
@@ -1,74 +1,53 @@
...
-  destroy(): void;
-}
-
-function setRootHistoryKey(
-  consumerLocation: history.Location,
-  rootHistoryKey: string | undefined
-): void {
-  consumerLocation.state = {...consumerLocation.state, rootHistoryKey};
-}
+export abstract class BaseHistory implements history.History {
```
Comment (clebert) [inline packages/history-service/src/base-history.ts]: Maybe rename this class to `ConsumerHistory`

File: packages/history-service/src/static-history.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    this.location = history.createLocation(pathOrLocation, state);
+    this.action = 'REPLACE';
+  }
+
+  public createHref(location: history.LocationDescriptorObject): history.Href {
+    return history.createPath(location);
+  }
+}
+
+export class StaticHistory extends BaseHistory {
```
Comment (clebert) [inline packages/history-service/src/static-history.ts]: rename `StaticConsumerHistory extends ConsumerHistory`

File: packages/history-service/src/browser-history.ts [hunk 13]
```diff
@@ -1,22 +1,33 @@
+import equal from 'fast-deep-equal';
 import * as history from 'history';
 import {BaseHistory} from './base-history';
 
 export class BrowserHistory extends BaseHistory {
```
Comment (clebert) [inline packages/history-service/src/browser-history.ts]: BrowserConsumerHistory

File: packages/history-service/src/browser-history.ts [hunk 14]
```diff
@@ -1,22 +1,33 @@
...
+  private readonly listeners = new Set<history.LocationListener>();
+  private readonly unregisterCallbacks: history.UnregisterCallback[] = [];
+
+  public destroy(): void {
+    this.unregisterCallbacks.forEach(unregister => unregister());
+    this.rootHistory.replace(this.createRootLocation(undefined));
```
Comment (clebert) [inline packages/history-service/src/browser-history.ts]: Please add a comment

Comment (clebert) [reply packages/history-service/src/browser-history.ts]: Maybe you could move this domain knowledge to its own multiplexer class

File: packages/history-service/src/history-multiplexer.ts [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    consumerId: string,
+    consumerLocation: history.Location
+  ): history.Href {
+    return this.rootHistory.createHref(
+      this.createRootLocation(consumerId, consumerLocation)
+    );
+  }
+
+  public getConsumerLocation(consumerId: string): history.Location {
+    const path =
```
Comment (clebert) [inline packages/history-service/src/history-multiplexer.ts]: maybe rename this `consumerPath`

File: packages/history-service/src/history-multiplexer.ts [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    consumerId: string,
+    consumerLocation: history.Location | undefined
+  ): history.Location {
+    const rootLocation = this.rootLocationTransformer.createRootLocation(
+      consumerLocation,
+      this.rootHistory.location,
+      consumerId
+    );
+
+    const state = this.rootHistory.location.state;
```
Comment (clebert) [inline packages/history-service/src/history-multiplexer.ts]: maybe rename this to `rootState`

File: packages/history-service/src/__tests__/index-node.test.ts [hunk 17]
```diff
@@ -0,0 +1,367 @@
+/**
+ * @jest-environment node
+ */
+
+// tslint:disable:no-non-null-assertion no-unbound-method
```
Comment (clebert) [inline packages/history-service/src/__tests__/index-node.test.ts]: newline after the tslint comment

File: packages/history-service/src/__tests__/index.test.ts [hunk 18]
```diff
@@ -1,16 +1,23 @@
 // tslint:disable:no-non-null-assertion no-unbound-method
```
Comment (clebert) [inline packages/history-service/src/__tests__/index.test.ts]: newline

**Pull request merged**